### PR TITLE
Log dropped telemetry item count on shutdown

### PR DESCRIPTION
**Pull request by mic-max**
This feature is important as we want to let users know somehow that we are dropping items.

## Changes
- On shutdown, write to an EventSource log with the count of dropped telemetry items since the start
- We may consider writing to the log the current number of dropped items every N drops. In this case, the first item dropped should write and then every subsequent Nth dropped item will write to the log.
  - i.e., The check of `this.droppedCount % N == 0` should happen on the line before `Interlocked.Increment(ref this.droppedCount);`

No dropped items:
image
Dropped items:
image

Note: I think the extra space at the end is being added by PerfView?

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by reyang — APPROVED** [review]
Need a changelog entry 😃

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 1]
```diff
@@ -186,6 +186,8 @@ protected override bool OnShutdown(int timeoutMilliseconds)
             this.shutdownDrainTarget = this.circularBuffer.AddedCount;
             this.shutdownTrigger.Set();
 
+            OpenTelemetrySdkEventSource.Log.DroppedExportProcessorItems(this.droppedCount);
```
Comment (cijothomas) [inline src/OpenTelemetry/BatchExportProcessor.cs:189]: Couple suggestion which I think will make this more useful:
1. As this is shared for LogRecord, Activity, we need to let use know which telemetry is dropped. Consider including tyeof(T) or similar.
2. If droppedCount is non-zero, we should be logging at higher severity than informational. We can either detect this here and call different methods, or in the actual EventSource class itself (by using a non-event, which inturn calls Info or Warning one based on droppedCount)

**Review by cijothomas — CHANGES_REQUESTED** [review]
I have left a comment about making this more useful for end user.

File: src/OpenTelemetry/BatchExportProcessor.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry/BatchExportProcessor.cs:189]: Perhaps also adding some indication which processor instance is dropping telemetry? (e.g. in case there are multiple batch exporting processors).

Comment (cijothomas) [reply src/OpenTelemetry/BatchExportProcessor.cs:189]: it may be hard.. I think we should be able to atleast include the nameof(exporter) this batchexportprocesor was exporting to.

Comment (reyang) [reply src/OpenTelemetry/BatchExportProcessor.cs:189]: Yeah, something like the name of the exporter and the instance id (or object hash).

E.g. "ZipkinExporter@0x12345678 dropped 100 items on the floor!".

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs [hunk 2]
```diff
@@ -95,7 +95,7 @@ public OtlpExporterOptions()
         /// <summary>
         /// Gets or sets the max waiting time (in milliseconds) for the backend to process each span batch. The default value is 10000.
         /// </summary>
-        public int TimeoutMilliseconds { get; set; } = 10000;
+        public uint TimeoutMilliseconds { get; set; } = 10000;
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:98]: This would be a breaking change.

**Review by reyang — CHANGES_REQUESTED** [review]
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2331/files#r707623666

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:98]: By convention all the .NET APIs use int as timeoutMs https://docs.microsoft.com/en-us/dotnet/api/system.threading.waithandle.waitone?view=net-5.0#System_Threading_WaitHandle_WaitOne_System_Int32_System_Boolean_.

Comment (mic-max) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:98]: Oops I forgot this branch was still active in a PR and while making sure I had everything from my desktop pushed before switching to my laptop for travel.
The last commits were just as the commit message says ideas 😅 I'll clean this up and make the suggested changes in the above comment as well.

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 3]
```diff
@@ -186,7 +186,7 @@ protected override bool OnShutdown(int timeoutMilliseconds)
             this.shutdownDrainTarget = this.circularBuffer.AddedCount;
             this.shutdownTrigger.Set();
 
-            OpenTelemetrySdkEventSource.Log.DroppedExportProcessorItems(this.droppedCount);
+            OpenTelemetrySdkEventSource.Log.DroppedExportProcessorItems(this.droppedCount, this.GetType());
```
Comment (reyang) [inline src/OpenTelemetry/BatchExportProcessor.cs:189]: Consider moving the `this.GetType()` call into the `if(IsEnabled)` block.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 4]
```diff
@@ -135,6 +135,22 @@ public void MissingPermissionsToReadEnvironmentVariable(SecurityException ex)
             }
         }
 
+        [NonEvent]
+        public void DroppedExportProcessorItems(long droppedCount, Type type)
+        {
+            if (this.IsEnabled(EventLevel.Informational, EventKeywords.All) && droppedCount <= 0)
```
Comment (reyang) [inline src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:141]: This might be faster:

```suggestion
            if (droppedCount <= 0 && this.IsEnabled(EventLevel.Informational, EventKeywords.All))
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpExporterOptionsTests.cs [hunk 5]
```diff
@@ -82,6 +82,9 @@ public void OtlpExporterOptions_SetterOverridesEnvironmentVariable()
             Environment.SetEnvironmentVariable(OtlpExporterOptions.HeadersEnvVarName, "A=2,B=3");
             Environment.SetEnvironmentVariable(OtlpExporterOptions.TimeoutEnvVarName, "2000");
 
+            // If your application is targeting .NET Core 3.1, and you are using an insecure (HTTP) endpoint, the following switch must be set before adding OtlpExporter.
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpExporterOptionsTests.cs]: Is this related to this PR?

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs (hunk 4 above)
Comment (mic-max) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:141]: My reasoning for doing in that order was just to have a similar look to the other functions, but ya I'll take this suggestion 👍

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpExporterOptionsTests.cs (hunk 5 above)
Comment (mic-max) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpExporterOptionsTests.cs]: No, accidentally pushed this.

File: src/OpenTelemetry/BatchExportProcessor.cs (hunk 3 above)
Comment (mic-max) [reply src/OpenTelemetry/BatchExportProcessor.cs:189]: True, otherwise it will do unnecessary work when there are 0 dropped items, hopefully the common scenario!

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 6]
```diff
@@ -186,6 +186,8 @@ protected override bool OnShutdown(int timeoutMilliseconds)
             this.shutdownDrainTarget = this.circularBuffer.AddedCount;
             this.shutdownTrigger.Set();
 
+            OpenTelemetrySdkEventSource.Log.DroppedExportProcessorItems(this.GetType().Name, this.droppedCount);
```
Comment (cijothomas) [inline src/OpenTelemetry/BatchExportProcessor.cs:189]: this is not the exporter name. this'd be the name of the batchexportprocessor. Exporter example is "OTLPExporter", "ZipkinExporter" etc.

Comment (cijothomas) [reply src/OpenTelemetry/BatchExportProcessor.cs:189]: "BatchExportProcessor<Activity> exporting to OTLP/Jaeger/Zipkin/ExporterName dropped N items due to buffer full." -- ~~ this is more helpful

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 7]
```diff
@@ -186,6 +186,8 @@ protected override bool OnShutdown(int timeoutMilliseconds)
             this.shutdownDrainTarget = this.circularBuffer.AddedCount;
             this.shutdownTrigger.Set();
 
+            OpenTelemetrySdkEventSource.Log.DroppedExportProcessorItems(nameof(BatchExportProcessor<T>), typeof(T).Name, this.exporter.GetType().Name, this.droppedCount);
```
Comment (cijothomas) [inline src/OpenTelemetry/BatchExportProcessor.cs:189]: if you just do "this.GetType().Name" and "this.exporter.GetType().Name", we get what we need.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 8]
```diff
@@ -309,6 +328,18 @@ public void MissingPermissionsToReadEnvironmentVariable(string exception)
             this.WriteEvent(30, exception);
         }
 
+        [Event(31, Message = "'{0}<{1}>' exporting to '{2}' dropped '0' items.", Level = EventLevel.Informational)]
```
Comment (cijothomas) [inline src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:331]: see this : https://github.com/open-telemetry/opentelemetry-dotnet/pull/2331/files#r711278336
if you do that, you won need to do this stitching here.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 9]
```diff
@@ -309,6 +328,18 @@ public void MissingPermissionsToReadEnvironmentVariable(string exception)
             this.WriteEvent(30, exception);
         }
 
+        [Event(31, Message = "'{0}' exporting to '{1}' dropped '0' items.", Level = EventLevel.Informational)]
+        public void NoDroppedExportProcessorItems(string exportProcessorName, string itemName, string exporterName)
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:332]: @mic-max Can you fix this? and update the description with a sample warning/information message and then we can merge this.
(for future, we might include hashid of the particular instance of the processor/exporters, but we can address it in future PRs.)

**Comment by cijothomas** [comment]
@mic-max Looks like changelog has conflicts that must be resolved before we can merge. Can you fix them ?

**Review by cijothomas — APPROVED** [review]
